Round scorebar to two decimal places

ZacWarham opened the pull request:
Makes the requested change to close #10905 - 2 decimal places chosen so that it matches the number of places used for points

![image](https://github.com/user-attachments/assets/56d04f58-e5ce-4ce0-b8b9-814c3194dec2)

This was just some simple practice for me so feel free to disregard if this is a change that is not wanted (or if there is a better way to do this)

nwalters512 commented on apps/prairielearn/src/components/Scorebar.html.ts:
```diff
@@ -15,13 +15,13 @@ export function Scorebar(
     >
       <div class="progress-bar bg-success" style="width: ${Math.floor(Math.min(100, score))}%">
-        ${score >= 50 ? `${Math.floor(score)}%` : ''}
+        ${score >= 50 ? `${Math.round(score * 100) / 100}%` : ''}
```
We'd still want to use `Math.floor` here. We _never_ want to give students the impression that they got more points than they actually did. For instance, if their actual score was 21.1777777%, we don't want to show 21.18%, as that would be a higher number than what they actually earned.

I think 2 decimal points is reasonable, as that's the same as when we show points information:

![Screenshot 2024-11-07 at 13 10 25](https://github.com/user-attachments/assets/9370b9a2-18aa-459f-87e4-3a960f11e915)

As long as we're here: consider lifting this computation up to a variable, like:

```js
const percentage = `${Math.round(score * 100) / 100}%`;
```

ZacWarham replied:
Good point - reverting back to floor now and moving it to a variable

mwest1066 commented on apps/prairielearn/src/components/Scorebar.html.ts:
```diff
@@ -9,19 +9,22 @@ export function Scorebar(
 ) {
   if (score == null) return '';
+
+  const percentage = `${Math.floor(score * 100) / 100}%`;
```
JavaScript actually has a special function for this, which also guarantees that there will be no issues with floating point representations. I'd also change the names here for clarity:
* Change the argument of the function from `score` to `scorePerc`
* Use a name for the string that indicates how it differs from the numeric value, e.g., `scorePercString`

This gives:
```suggestion
  const scorePercString = `${scorePerc.toFixed(2)}%`;
```

ZacWarham replied:
Doesn't toFixed() round up or down but we only want to round down?

Eg. `${scorePerc.toFixed(1)}%` rounds to 11.3 not 11.2

![image](https://github.com/user-attachments/assets/2b0944cc-94af-4a14-a84c-9e6a0a487238)

jonatanschroeder replied:
I wouldn't use toFixed because it forces zeros to the right (54.00). You can use formatPoints instead.

jonatanschroeder replied:
https://github.com/PrairieLearn/PrairieLearn/blob/dcfa464e261a0847dfc3803dae180e790152a6d9/apps/prairielearn/src/lib/format.ts#L9

mwest1066 replied:
> Doesn't toFixed() round up or down but we only want to round down?

Excellent point, I thought about this and I meant to leave in the `* 100 / 100` to handle this, but then I forgot!

But @jonatanschroeder's suggestion of `formatPoints()` is of course the correct answer here.

mwest1066 commented:
As I commented on https://github.com/PrairieLearn/PrairieLearn/issues/10905 it's not clear to me whether two decimal places are needed. That feels like a lot of precision (1 in 10000)? Could we get away with one decimal place?

ZacWarham commented:
I have updated the variables names as suggested by @mwest1066 and switched to `formatPoints()` as suggested by @jonatanschroeder 

I have provided examples below of it with both 2dp and 1dp since there seems to be mixed opinions on which it should be (currently using the `formatPoints` default of 2)

> https://github.com/PrairieLearn/PrairieLearn/blob/dcfa464e261a0847dfc3803dae180e790152a6d9/apps/prairielearn/src/lib/format.ts#L9

Current: 2dp - matching points rounding;

![2dp eg1](https://github.com/user-attachments/assets/8e706ce5-ff7d-42cc-b6f1-dd4a93cc744c)
![2dp eg2](https://github.com/user-attachments/assets/021bb58a-3e69-4ef5-a6f5-d37522db1c75)

Alternative: 1dp - reducing clutter;

![1dp eg1](https://github.com/user-attachments/assets/d79e815c-3e3b-4ac3-9292-e0d991fb25e9)
![1dp eg2](https://github.com/user-attachments/assets/4a15b680-0c5e-4c70-8fb4-4effda5ab448)

mwest1066 commented:
I'm continuing the discussion about this PR in the comments for #10905.

nwalters512 commented:
As discussed on #10905, we're going to resolve students' need for higher-precision score information a different way.
